doc/source/dev/toolchain.rst

```diff
@@ -0,0 +1,136 @@
+Toolchain and Roadmap
+=====================
+SciPy requires several tools in order to be built.
+
+- First and foremost, SciPy is a Python project hence it requires a Python environment.
+- The LAPACK and OpenBLAS numerical libraries need to be installed.
+- Compilers for C, Cython and Fortran code are needed.
+- The Python environment needs the ``NumPy`` package to be installed.
+- Testing requires the ``pytest`` Python package.
+- Building the documentation requires Sphinx and ``numpydoc``.
+
+The tooling used to build CPython has some implications for the tooling used in building SciPy.
+
+Python Versions
+---------------
+
+SciPy is compatible with several versions of Python.
+[*NOTE: The level of support and dates is currently under discussion.
+This table shows what it could look like.*]
+
+================  =================================================
+ Date             Pythons supported
+================  =================================================
+ <= 2018           Py2.7, Py3.5 - Py3.7
+ 2019              Py3.5 - (but Py2.7-specific code not removed)
+ 2020              Py3.?- (removal of Py2.7-specific code permitted)
+================  =================================================
+
+
+C Compilers
+-----------
+SciPy is compatible with most modern C compilers.  However CPython on Windows is
+built with specific versions of the Microsoft Visual C++ compiler [1]_, as is the
+corresponding build of SciPy.  This has implications for the C language standards
+that can be supported.
+
+===================   ==============   ===================
+CPython               MS Visual C++    C Standard
+===================   ==============   ===================
+3.5, 3.6                14.0           C90 + most of C99
+3.3, 3.4                10.0           C90 + some of C99
+2.7, 3.0, 3.1, 3.2       9.0           C90
+===================   ==============   ===================
+
+
+C Language Standards
+--------------------
+[*NOTE: I'm unaware of any official SciPy decisions on the support of C standards.
+This table shows what it could look like.*]
```

Comment:
There's no official decision anywhere. It has been C89 forever, because that was the only option given the Windows situation.

It's tricky to make such decisions in advance (roadmap style). The thing is that we're evaluating such things by the impact on users and developers. Which is usually a case-by-case thing. E.g. when we decided to not ship without SSE2 support anymore, we just needed data on how many users as % of total user base was still running on older hardware without SSE2 support.

That said, document what we would like to see is very useful - let's just not call it a decision.

Reply:
https://github.com/scipy/scipy/pull/9915#issuecomment-475326101 has a suggestion on C++ standard as well

Reply:
I added a C++ standard statement, and some brief language to the effect that these are less "official" decisions and more "guidelines," especially for future adoptions of more modern standards